[完了] タスクの[完了/未完了]切り替え機能

mikiya1130 opened the pull request:
- タスクの[完了/未完了]を切り替えるためのチェックボックスを作成
- 完了したタスクは打ち消し線を引いて色を薄くする
- 完了したタスクを非表示にするためのチェックボックスを作成

close #19

mikiya1130 committed:
refactor: 余計な差分を防止するため

src ディレクトリ下のファイルにフォーマットを適用

mikiya1130 committed:
feat: タスク一覧を表示するため

mikiya1130 committed:
refactor: デフォルトのファイルが不必要になったため

mikiya1130 committed:
feat: 長いタスク名に対応するため

3点リーダーで省略して表示するように変更

mikiya1130 committed:
feat: 新たに追加するタスクを入力するため

Input field を追加

mikiya1130 committed:
chore: ボタンアイコンに使用するため

$ yarn add @chakra-ui/icons を実行

mikiya1130 committed:
feat: タスクを追加するイベントを発生させるため

タスクを追加するためのボタンを追加

mikiya1130 committed:
feat: タスクを登録するため

Enter キー押下、もしくは CheckIcon クリックでタスク登録のイベント発火

mikiya1130 committed:
feat: タスクの内容を取得するため

useRef を使用して input に入力した内容を取得

mikiya1130 committed:
refactor: タスク追加以外の処理も行う(計画)ため

タスク登録イベント発火時に呼ばれる関数の関数名を変更

mikiya1130 committed:
refactor: 空文字の時はタスク登録処理を行わないようにするため

空文字チェックを追加

mikiya1130 committed:
feat: タスクの追加を反映させるため

- tasks list を useState で書き換え
- addTask 関数を実装

mikiya1130 committed:
refactor: タスクの追加処理が肥大化したため

タスクの追加フィールドと追加処理を別ファイルに分離

mikiya1130 committed:
feat: タスク追加後フィールドを空にするため

resetField 関数を実装
空文字で上書きできるようにするため、Input Field の値を useState での管理に変更

mikiya1130 committed:
refactor: 'Task' をコンポーネントとして定義するため

interface 'Task' を 'iTask' に修正

mikiya1130 committed:
refactor: タスクの編集機能を追加するのでコンポーネントを分割しておくべきだと考えたため

1つのタスク要素に相当するコンポーネントを別ファイルに分離

mikiya1130 committed:
feat: タスクを編集できるようにするため

タスクの編集機能を追加

mikiya1130 committed:
feat: タスクを一意に識別するため

iTask の key: create にタスク作成時の timestamp を設定

mikiya1130 committed:
feat: task 編集時に tasks 配列の値を更新し表示内容と同期させておくため

Editable コンポーネントを用いたタスクの編集完了時にコールバック関数を呼び出し、tasks 配列を更新する

mikiya1130 committed:
refactor: 不要になったため

mikiya1130 committed:
chore: vscode で indent を space x2 に変更するため

mikiya1130 committed:
style: 「追加」のイメージにより近いため

タスクの追加ボタンのアイコンを変更(✔ → ＋)

mikiya1130 committed:
feat: タスクの[完了/未完了]を識別するため

isDone フラグを追加

mikiya1130 committed:
feat:  チェックの状態と配列の isDone フラグの状態を同期させるため

mikiya1130 committed:
feat: タスクの完了を分かりやすくするため

完了したタスクは打ち消し線を引いて色を薄くする

mikiya1130 committed:
feat: 完了したタスクを非表示にできるようにするため

mikiya1130 commented on todo-app/src/Task.tsx:
```diff
@@ -84,8 +76,40 @@ const Task = ({ task, updateTasks }: Props) => {
+    );
+  };
+
+  if (task.isDone) {
+    return (
+      <Editable
+        p={3}
+        m={1}
+        w="100%"
+        color="gray"
+        bg="blue.100"
+        defaultValue={task.text}
+        as="s"
+        isPreviewFocusable={false}
+        onSubmit={(nextValue) => onSubmit(nextValue)}
+      >
+        <TaskContents />
+      </Editable>
+    );
+  } else {
+    return (
+      <Editable
+        p={3}
+        m={1}
+        w="100%"
+        bg="blue.100"
+        defaultValue={task.text}
+        isPreviewFocusable={false}
+        onSubmit={(nextValue) => onSubmit(nextValue)}
+      >
+        <TaskContents />
+      </Editable>
+    );
+  }
```
この部分の適切な書き方を教えていただきたいです。
タスクが完了したら打ち消し線を引いて文字色を灰色にしている箇所です。

やりたいこととしては、コンポーネントに適用するスタイルの、特定の条件に応じた変更です。
ここでは、`task.isDone: boolean`の状態に応じて、Editable コンポーネントに`color="gray"`と`as="s"`を適用するか否かが変化しています。

watagit replied:
私だったら、別コンポーネントとして`Editable`をカスタマイズした`CustomEditable`を作るかもです。

例
`CustomEditable.tsx`
```ts
import { Editable, EditableProps } from "@chakra-ui/editable";

type CustomProps = {
  isDone: boolean;
};

type Props = EditableProps & CustomProps;

// props として受け取った isDone によって表示を変えるコンポーネント
const CustomEditable = (props: Props) => {
  const { isDone } = props;

  return isDone ? (
    <Editable color="gray" as="s" {...props} />  // ここで color と as を指定しちゃう
  ) : (
    <Editable {...props} />
  );
};

export default CustomEditable;
```
これを作ると何が嬉しいかというと、以下のように、`Task.tsx`で考えることが減ります。

`Task.tsx`
```ts
import CustomEditable from "./CustomEditable";

// ~~~

const Task = ({ task, updateTasks }: Props) => {

  // ~~~

  // CustomEditable に isDone を渡すだけで良くなる
  return (
    <CustomEditable
      isDone={task.isDone}  // ✨ 
      p={3}
      m={1}
      w="100%"
      bg="blue.100"
      defaultValue={task.text}
      isPreviewFocusable={false}
      onSubmit={(nextValue) => onSubmit(nextValue)}
    >
      <TaskContents />
    </CustomEditable>
  );
};
```

mikiya1130 replied:
ご丁寧にありがとうございます。
継承的な(React ではそう言わないかもやけど)書き方すれば良いのですね。
参考になりました。
intersection 型や Props の展開なども、React初心者としては学びの多いコードでした。

mikiya1130 committed:
refactor: コンポーネントに与える属性の条件分岐を簡潔に行うため

カスタムコンポーネントの作成
https://github.com/bislab-temporary/react-study-morisaki/pull/20#discussion_r719127767 を参考

mikiya1130 committed:
refactor: コンポーネントの切り出しが不要なため

CustomEditable の使用により TaskContents コンポーネントの呼び出しが1箇所になったため
